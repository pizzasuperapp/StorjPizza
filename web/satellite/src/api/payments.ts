// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { ErrorUnauthorized } from '@/api/errors/ErrorUnauthorized';
import {
    AccountBalance,
    Coupon,
    CreditCard,
    PaymentsApi,
    PaymentsHistoryItem,
    ProjectUsageAndCharges,
    TokenDeposit
} from '@/types/payments';
import { HttpClient } from '@/utils/httpClient';
import { Time } from '@/utils/time';
import { ErrorTooManyRequests } from './errors/ErrorTooManyRequests';

/**
 * PaymentsHttpApi is a http implementation of Payments API.
 * Exposes all payments-related functionality
 */
export class PaymentsHttpApi implements PaymentsApi {
    private readonly client: HttpClient = new HttpClient();
    private readonly ROOT_PATH: string = '/api/v0/payments';

    /**
     * Get account balance.
     *
     * @returns balance in cents
     * @throws Error
     */
    public async getBalance(): Promise<AccountBalance> {
        const path = `${this.ROOT_PATH}/account/balance`;
        const response = await this.client.get(path);

        if (!response.ok) {
            if (response.status === 401) {
                throw new ErrorUnauthorized();
            }

            throw new Error('Can not get account balance');
        }

        const balance = await response.json();
        if (balance) {
            return new AccountBalance(balance.freeCredits, balance.coins);
        }

        return new AccountBalance();
    }

    /**
     * Try to set up a payment account.
     *
     * @throws Error
     */
    public async setupAccount(): Promise<string> {
        const path = `${this.ROOT_PATH}/account`;
        const response = await this.client.post(path, null);
        const couponType = await response.json();

        if (response.ok) {
            return couponType;
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not setup account');
    }

    /**
     * projectsUsageAndCharges returns usage and how much money current user will be charged for each project which he owns.
     */
    public async projectsUsageAndCharges(start: Date, end: Date): Promise<ProjectUsageAndCharges[]> {
        const since = Time.toUnixTimestamp(start).toString();
        const before = Time.toUnixTimestamp(end).toString();
        const path = `${this.ROOT_PATH}/account/charges?from=${since}&to=${before}`;
        const response = await this.client.get(path);

        if (!response.ok) {
            if (response.status === 401) {
                throw new ErrorUnauthorized();
            }

            throw new Error('can not get projects charges');
        }

        const charges = await response.json();
        if (charges) {
            return charges.map(charge =>
                new ProjectUsageAndCharges(
                    new Date(charge.since),
                    new Date(charge.before),
                    charge.egress,
                    charge.storage,
                    charge.segmentCount,
                    charge.projectId,
                    charge.storagePrice,
                    charge.egressPrice,
                    charge.segmentPrice,
                ),
            );
        }

        return [];
    }

    /**
     * Add credit card.
     *
     * @param token - stripe token used to add a credit card as a payment method
     * @throws Error
     */
    public async addCreditCard(token: string): Promise<void> {
        const path = `${this.ROOT_PATH}/cards`;
        const response = await this.client.post(path, token);

        if (response.ok) {
            return;
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not add credit card');
    }

    /**
     * Detach credit card from payment account.
     *
     * @param cardId
     * @throws Error
     */
    public async removeCreditCard(cardId: string): Promise<void> {
        const path = `${this.ROOT_PATH}/cards/${cardId}`;
        const response = await this.client.delete(path);

        if (response.ok) {
            return;
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not remove credit card');
    }

    /**
     * Get list of user`s credit cards.
     *
     * @returns list of credit cards
     * @throws Error
     */
    public async listCreditCards(): Promise<CreditCard[]> {
        const path = `${this.ROOT_PATH}/cards`;
        const response = await this.client.get(path);

        if (!response.ok) {
            if (response.status === 401) {
                throw new ErrorUnauthorized();
            }
            throw new Error('can not list credit cards');
        }

        const creditCards = await response.json();

        if (creditCards) {
            return creditCards.map(card => new CreditCard(card.id, card.expMonth, card.expYear, card.brand, card.last4, card.isDefault));
        }

        return [];
    }

    /**
     * Make credit card default.
     *
     * @param cardId
     * @throws Error
     */
    public async makeCreditCardDefault(cardId: string): Promise<void> {
        const path = `${this.ROOT_PATH}/cards`;
        const response = await this.client.patch(path, cardId);

        if (response.ok) {
            return;
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not make credit card default');
    }

    /**
     * Returns a list of invoices, transactions and all others payments history items for payment account.
     *
     * @returns list of payments history items
     * @throws Error
     */
    public async paymentsHistory(): Promise<PaymentsHistoryItem[]> {
        const path = `${this.ROOT_PATH}/billing-history`;
        const response = await this.client.get(path);

        if (!response.ok) {
            if (response.status === 401) {
                throw new ErrorUnauthorized();
            }
            throw new Error('can not list billing history');
        }

        const paymentsHistoryItems = await response.json();
        if (paymentsHistoryItems) {
            return paymentsHistoryItems.map(item =>
                new PaymentsHistoryItem(
                    item.id,
                    item.description,
                    item.amount,
                    item.received,
                    item.status,
                    item.link,
                    new Date(item.start),
                    new Date(item.end),
                    item.type,
                    item.remaining,
                ),
            );
        }

        return [];
    }

    /**
     * makeTokenDeposit process coin payments.
     *
     * @param amount
     * @throws Error
     */
    public async makeTokenDeposit(amount: number): Promise<TokenDeposit> {
        const path = `${this.ROOT_PATH}/tokens/deposit`;
        const response = await this.client.post(path, JSON.stringify({ amount }));

        if (!response.ok) {
            if (response.status === 401) {
                throw new ErrorUnauthorized();
            }

            throw new Error('can not process coin payment');
        }

        const result = await response.json();

        return new TokenDeposit(result.amount, result.address, result.link);
    }

    /**
     * applyCouponCode applies a coupon code.
     *
     * @param couponCode
     * @throws Error
     */
    public async applyCouponCode(couponCode: string): Promise<Coupon> {
        const path = `${this.ROOT_PATH}/coupon/apply`;
        const response = await this.client.patch(path, couponCode);
        const errMsg = `Could not apply coupon code "${couponCode}"`;

        if (response.ok) {
            const coupon = await response.json();

            if (!coupon) {
                throw new Error(errMsg);
            }

            return new Coupon(
                coupon.id,
                coupon.promoCode,
                coupon.name,
                coupon.amountOff,
                coupon.percentOff,
                new Date(coupon.addedAt),
                coupon.expiresAt ? new Date(coupon.expiresAt) : null,
                coupon.duration
            );
        }

        switch (response.status) {
        case 429:
            throw new ErrorTooManyRequests('You\'ve exceeded limit of attempts, try again in 5 minutes');
        case 401:
            throw new ErrorUnauthorized(errMsg);
        default:
            throw new Error(errMsg);
        }
    }

    /**
     * getCoupon returns the coupon applied to the user.
     *
     * @throws Error
     */
    public async getCoupon(): Promise<Coupon | null> {
        const path = `${this.ROOT_PATH}/coupon`;
        const response = await this.client.get(path);
        if (!response.ok) {
            if (response.status === 401) {
                throw new ErrorUnauthorized();
            }

            throw new Error('cannot retrieve coupon');
        }

        const coupon = await response.json();

        if (!coupon) {
            return null;
        }

        return new Coupon(
            coupon.id,
            coupon.promoCode,
            coupon.name,
            coupon.amountOff,
            coupon.percentOff,
            new Date(coupon.addedAt),
            coupon.expiresAt ? new Date(coupon.expiresAt) : null,
            coupon.duration
        );
    }
}
