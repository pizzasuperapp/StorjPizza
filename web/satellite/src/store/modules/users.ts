// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { DisableMFARequest, UpdatedUser, User, UsersApi } from '@/types/users';
import { MetaUtils } from '@/utils/meta';
import {StoreModule} from "@/types/store";

export const USER_ACTIONS = {
    UPDATE: 'updateUser',
    GET: 'getUser',
    ENABLE_USER_MFA: 'enableUserMFA',
    DISABLE_USER_MFA: 'disableUserMFA',
    GENERATE_USER_MFA_SECRET: 'generateUserMFASecret',
    GENERATE_USER_MFA_RECOVERY_CODES: 'generateUserMFARecoveryCodes',
    CLEAR: 'clearUser',
};

export const USER_MUTATIONS = {
    SET_USER: 'setUser',
    SET_USER_MFA_SECRET: 'setUserMFASecret',
    SET_USER_MFA_RECOVERY_CODES: 'setUserMFARecoveryCodes',
    UPDATE_USER: 'updateUser',
    CLEAR: 'clearUser',
};

export class UsersState {
    public user: User = new User();
    public userMFASecret = '';
    public userMFARecoveryCodes: string[] = [];
}

const {
    GET,
    UPDATE,
    ENABLE_USER_MFA,
    DISABLE_USER_MFA,
    GENERATE_USER_MFA_SECRET,
    GENERATE_USER_MFA_RECOVERY_CODES,
} = USER_ACTIONS;

const {
    SET_USER,
    UPDATE_USER,
    SET_USER_MFA_SECRET,
    SET_USER_MFA_RECOVERY_CODES,
    CLEAR,
} = USER_MUTATIONS;

interface UsersContext {
    state: UsersState
    commit: (string, ...unknown) => void
}

/**
 * creates users module with all dependencies
 *
 * @param api - users api
 */
export function makeUsersModule(api: UsersApi): StoreModule<UsersState, UsersContext> {
    return {
        state: new UsersState(),

        mutations: {
            [SET_USER](state: UsersState, user: User): void {
                state.user = user;

                if (user.projectLimit === 0) {
                    const limitFromConfig = MetaUtils.getMetaContent('default-project-limit');

                    state.user.projectLimit = parseInt(limitFromConfig);

                    return;
                }

                state.user.projectLimit = user.projectLimit;
            },
            [CLEAR](state: UsersState): void {
                state.user = new User();
                state.user.projectLimit = 1;
            },
            [UPDATE_USER](state: UsersState, user: UpdatedUser): void {
                state.user.fullName = user.fullName;
                state.user.shortName = user.shortName;
            },
            [SET_USER_MFA_SECRET](state: UsersState, secret: string): void {
                state.userMFASecret = secret;
            },
            [SET_USER_MFA_RECOVERY_CODES](state: UsersState, codes: string[]): void {
                state.userMFARecoveryCodes = codes;
                state.user.mfaRecoveryCodeCount = codes.length;
            },
        },

        actions: {
            [UPDATE]: async function ({commit}: UsersContext, userInfo: UpdatedUser): Promise<void> {
                await api.update(userInfo);

                commit(UPDATE_USER, userInfo);
            },
            [GET]: async function ({commit}: UsersContext): Promise<User> {
                const user = await api.get();

                commit(SET_USER, user);

                return user;
            },
            [DISABLE_USER_MFA]: async function (_, request: DisableMFARequest): Promise<void> {
                await api.disableUserMFA(request.passcode, request.recoveryCode);
            },
            [ENABLE_USER_MFA]: async function (_, passcode: string): Promise<void> {
                await api.enableUserMFA(passcode);
            },
            [GENERATE_USER_MFA_SECRET]: async function ({commit}: UsersContext): Promise<void> {
                const secret = await api.generateUserMFASecret();

                commit(SET_USER_MFA_SECRET, secret);
            },
            [GENERATE_USER_MFA_RECOVERY_CODES]: async function ({commit}: UsersContext): Promise<void> {
                const codes = await api.generateUserMFARecoveryCodes();

                commit(SET_USER_MFA_RECOVERY_CODES, codes);
            },
            [CLEAR]: function({commit}: UsersContext) {
                commit(CLEAR);
            },
        },

        getters: {
            user: (state: UsersState): User => state.user,
            userName: (state: UsersState): string => state.user.getFullName(),
        },
    };
}
