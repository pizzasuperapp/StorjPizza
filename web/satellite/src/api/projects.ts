// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BaseGql } from '@/api/baseGql';
import { ErrorUnauthorized } from '@/api/errors/ErrorUnauthorized';
import {
    DataStamp,
    Project,
    ProjectFields,
    ProjectLimits,
    ProjectsApi,
    ProjectsCursor,
    ProjectsPage,
    ProjectsStorageBandwidthDaily,
} from '@/types/projects';
import { HttpClient } from '@/utils/httpClient';
import { Time } from "@/utils/time";

export class ProjectsApiGql extends BaseGql implements ProjectsApi {
    private readonly http: HttpClient = new HttpClient();
    private readonly ROOT_PATH: string = '/api/v0/projects';

    /**
     * Creates project.
     *
     * @param projectFields - contains project information
     * @throws Error
     */
    public async create(projectFields: ProjectFields): Promise<Project> {
        const query =
            `mutation($name: String!, $description: String!) {
                createProject(
                    input: {
                        name: $name,
                        description: $description,
                    }
                ) {id}
            }`;

        const variables = {
            name: projectFields.name,
            description: projectFields.description,
        };

        const response = await this.mutate(query, variables);

        return new Project(response.data.createProject.id, variables.name, variables.description, '', projectFields.ownerId);
    }

    /**
     * Fetch projects.
     *
     * @returns Project[]
     * @throws Error
     */
    public async get(): Promise<Project[]> {
        const query = `query {
            myProjects{
                name
                id
                description
                createdAt
                ownerId
            }
        }`;

        const response = await this.query(query);

        return response.data.myProjects.map((project: Project) => {
            return new Project(
                project.id,
                project.name,
                project.description,
                project.createdAt,
                project.ownerId,
            );
        });
    }

    /**
     * Update project name and description.
     *
     * @param projectId - project ID
     * @param projectFields - project fields
     * @param projectLimits - project limits
     * @returns Project[]
     * @throws Error
     */
    public async update(projectId: string, projectFields: ProjectFields, projectLimits: ProjectLimits): Promise<void> {
        const query =
            `mutation($projectId: String!, $name: String!, $description: String!, $storageLimit: String!, $bandwidthLimit: String!) {
                updateProject(
                    id: $projectId,
                    projectFields: {
                        name: $name,
                        description: $description,
                    },
                    projectLimits: {
                        storageLimit: $storageLimit,
                        bandwidthLimit: $bandwidthLimit,
                    }
                ) {name}
            }`;

        const variables = {
            projectId: projectId,
            name: projectFields.name,
            description: projectFields.description,
            storageLimit: projectLimits.storageLimit.toString(),
            bandwidthLimit: projectLimits.bandwidthLimit.toString(),
        };

        await this.mutate(query, variables);
    }

    /**
     * Delete project.
     *
     * @param projectId - project ID
     * @throws Error
     */
    public async delete(projectId: string): Promise<void> {
        const query =
            `mutation($projectId: String!) {
                deleteProject(
                    id: $projectId
                ) {name}
            }`;

        const variables = {
            projectId: projectId,
        };

        await this.mutate(query, variables);
    }

    /**
     * Get project limits.
     *
     * @param projectId- project ID
     * throws Error
     */
    public async getLimits(projectId: string): Promise<ProjectLimits> {
        const path = `${this.ROOT_PATH}/${projectId}/usage-limits`;
        const response = await this.http.get(path);

        if (response.ok) {
            const limits = await response.json();

            return new ProjectLimits(
                limits.bandwidthLimit,
                limits.bandwidthUsed,
                limits.storageLimit,
                limits.storageUsed,
                limits.objectCount,
                limits.segmentCount,
            );
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not get usage limits');
    }

    /**
     * Get total limits for all the projects that user owns.
     *
     * throws Error
     */
    public async getTotalLimits(): Promise<ProjectLimits> {
        const path = `${this.ROOT_PATH}/usage-limits`;
        const response = await this.http.get(path);

        if (response.ok) {
            const limits = await response.json();

            return new ProjectLimits(
                limits.bandwidthLimit,
                limits.bandwidthUsed,
                limits.storageLimit,
                limits.storageUsed,
            );
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not get total usage limits');
    }

    /**
     * Get project daily usage for specific date range.
     *
     * @param projectId- project ID
     * @param start- since date
     * @param end- before date
     * throws Error
     */
    public async getDailyUsage(projectId: string, start: Date, end: Date): Promise<ProjectsStorageBandwidthDaily> {
        const since = Time.toUnixTimestamp(start).toString();
        const before = Time.toUnixTimestamp(end).toString();
        const path = `${this.ROOT_PATH}/${projectId}/daily-usage?from=${since}&to=${before}`;
        const response = await this.http.get(path);

        if (response.ok) {
            const usage = await response.json();

            return new ProjectsStorageBandwidthDaily(
                usage.storageUsage.map(el => {
                    const date = new Date(el.date)
                    date.setHours(0, 0, 0, 0)
                    return new DataStamp(el.value, date)
                }),
                usage.allocatedBandwidthUsage.map(el => {
                    const date = new Date(el.date)
                    date.setHours(0, 0, 0, 0)
                    return new DataStamp(el.value, date)
                }),
                usage.settledBandwidthUsage.map(el => {
                    const date = new Date(el.date)
                    date.setHours(0, 0, 0, 0)
                    return new DataStamp(el.value, date)
                }),
            );
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not get project daily usage');
    }

    /**
     * Fetch owned projects.
     *
     * @returns ProjectsPage
     * @throws Error
     */
    public async getOwnedProjects(cursor: ProjectsCursor): Promise<ProjectsPage> {
        const query =
            `query($limit: Int!, $page: Int!) {
                ownedProjects( cursor: { limit: $limit, page: $page } ) {
                    projects {
                        id,
                        name,
                        ownerId,
                        description,
                        createdAt,
                        memberCount
                    },
                    limit,
                    offset,
                    pageCount,
                    currentPage,
                    totalCount
                 }
             }`;

        const variables = {
            limit: cursor.limit,
            page: cursor.page,
        };

        const response = await this.query(query, variables);

        return this.getProjectsPage(response.data.ownedProjects);
    }

    /**
     * Method for mapping projects page from json to ProjectsPage type.
     *
     * @param page anonymous object from json
     */
    private getProjectsPage(page: any): ProjectsPage { // eslint-disable-line @typescript-eslint/no-explicit-any
        if (!page) {
            return new ProjectsPage();
        }

        const projects: Project[] = page.projects.map(key =>
            new Project(
                key.id,
                key.name,
                key.description,
                key.createdAt,
                key.ownerId,
                false,
                key.memberCount));

        return new ProjectsPage(projects, page.limit, page.offset, page.pageCount, page.currentPage, page.totalCount);
    }

}
