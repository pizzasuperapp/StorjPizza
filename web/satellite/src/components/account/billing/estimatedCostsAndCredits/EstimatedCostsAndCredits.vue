// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="current-month-area">
        <VLoader v-if="isDataFetching" class="consts-loader" />
        <template v-else>
            <h1 class="current-month-area__costs">{{ priceSummary | centsToDollars }}</h1>
            <span class="current-month-area__title">Estimated Charges for {{ chosenPeriod }}</span>
            <p class="current-month-area__info">
                If you still have Storage and Bandwidth remaining in your free tier, you wonâ€™t be charged. This information
                is to help you estimate what charges would have been had you graduated to the paid tier.
            </p>
            <div class="current-month-area__content">
                <p class="current-month-area__content__title">DETAILS</p>
                <UsageAndChargesItem
                    v-for="usageAndCharges in projectUsageAndCharges"
                    :key="usageAndCharges.projectId"
                    :item="usageAndCharges"
                    class="item"
                />
            </div>
        </template>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import UsageAndChargesItem from '@/components/account/billing/estimatedCostsAndCredits/UsageAndChargesItem.vue';
import VLoader from '@/components/common/VLoader.vue';

import { PAYMENTS_ACTIONS } from '@/store/modules/payments';
import { PROJECTS_ACTIONS } from '@/store/modules/projects';
import { ProjectUsageAndCharges } from '@/types/payments';
import { MONTHS_NAMES } from '@/utils/constants/date';

// @vue/component
@Component({
    components: {
        UsageAndChargesItem,
        VLoader,
    },
})
export default class EstimatedCostsAndCredits extends Vue {
    public isDataFetching = true;

    /**
     * Lifecycle hook after initial render.
     * Fetches projects and usage rollup.
     */
    public async mounted(): Promise<void> {
        try {
            await this.$store.dispatch(PROJECTS_ACTIONS.FETCH);
        } catch (error) {
            this.isDataFetching = false;
            return;
        }

        try {
            await this.$store.dispatch(PAYMENTS_ACTIONS.GET_PROJECT_USAGE_AND_CHARGES_CURRENT_ROLLUP);

            this.isDataFetching = false;
        } catch (error) {
            await this.$notify.error(error.message);
        }
    }

    /**
     * projectUsageAndCharges is an array of all stored ProjectUsageAndCharges.
     */
    public get projectUsageAndCharges(): ProjectUsageAndCharges[] {
        return this.$store.state.paymentsModule.usageAndCharges;
    }

    /**
     * priceSummary returns price summary of usages for all the projects.
     */
    public get priceSummary(): number {
        return this.$store.state.paymentsModule.priceSummary;
    }

    /**
     * chosenPeriod returns billing period chosen by user.
     */
    public get chosenPeriod(): string {
        const dateFromStore = this.$store.state.paymentsModule.startDate;

        return `${MONTHS_NAMES[dateFromStore.getUTCMonth()]} ${dateFromStore.getUTCFullYear()}`;
    }
}
</script>

<style scoped lang="scss">
    h1,
    h2,
    p,
    span {
        margin: 0;
        color: #354049;
    }

    .current-month-area {
        margin-bottom: 32px;
        padding: 40px 40px 0;
        background-color: #fff;
        border-radius: 8px;
        font-family: 'font_regular', sans-serif;

        &__costs {
            font-size: 36px;
            line-height: 53px;
            color: #384b65;
            font-family: 'font_medium', sans-serif;
        }

        &__title {
            font-size: 16px;
            line-height: 24px;
            color: #909090;
        }

        &__info {
            font-size: 14px;
            line-height: 20px;
            color: #909090;
            margin: 15px 0 0;
        }

        &__content {
            margin-top: 35px;

            &__title {
                font-size: 16px;
                line-height: 23px;
                letter-spacing: 0.04em;
                text-transform: uppercase;
                color: #919191;
                margin-bottom: 25px;
            }

            &__usage-charges {
                margin: 18px 0 0;
                background-color: #f5f6fa;
                border-radius: 12px;
                cursor: pointer;
            }
        }
    }

    .item {
        border-top: 1px solid #c7cdd2;
    }

    .consts-loader {
        padding-bottom: 40px;
    }
</style>
