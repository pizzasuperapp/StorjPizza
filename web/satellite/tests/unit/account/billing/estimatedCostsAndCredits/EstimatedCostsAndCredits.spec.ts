// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';

import EstimatedCostsAndCredits from '@/components/account/billing/estimatedCostsAndCredits/EstimatedCostsAndCredits.vue';

import { makePaymentsModule, PAYMENTS_MUTATIONS } from '@/store/modules/payments';
import { makeProjectsModule, PROJECTS_MUTATIONS } from '@/store/modules/projects';
import { makeUsersModule, USER_MUTATIONS } from '@/store/modules/users';
import { ProjectUsageAndCharges } from '@/types/payments';
import { Project } from '@/types/projects';
import { User } from '@/types/users';
import { createLocalVue, mount } from '@vue/test-utils';

import { PaymentsMock } from '../../../mock/api/payments';
import { ProjectsApiMock } from '../../../mock/api/projects';
import { UsersApiMock } from '../../../mock/api/users';

const localVue = createLocalVue();
localVue.filter('centsToDollars', (cents: number): string => {
    return `USD $${(cents / 100).toFixed(2)}`;
});
localVue.use(Vuex);

const usersApi = new UsersApiMock();
const usersModule = makeUsersModule(usersApi);
const projectsApi = new ProjectsApiMock();
const projectsModule = makeProjectsModule(projectsApi);
const paymentsApi = new PaymentsMock();
const paymentsModule = makePaymentsModule(paymentsApi);
const store = new Vuex.Store({ modules: { usersModule, projectsModule, paymentsModule }});

const now = new Date(2020, 4, 1, 12 /* avoid timezone issues */, 1, 1, 1);
paymentsModule.state.startDate = now; // TODO: we shouldn't need to do this
paymentsModule.state.endDate = now;

const project = new Project('id', 'projectName', 'projectDescription', 'test', 'testOwnerId', true);
const project1 = new Project('id1', 'projectName1', 'projectDescription1', 'test', 'testOwnerId1', false);
const user = new User('testOwnerId');
const date = new Date(1970, 1, 1);
const projectCharge = new ProjectUsageAndCharges(date, date, 100, 100, 100, 'id', 100, 100, 100);
const projectCharge1 = new ProjectUsageAndCharges(date, date, 100, 100, 100, 'id1', 100, 100, 100);
const {
    SET_PROJECT_USAGE_AND_CHARGES,
    SET_PRICE_SUMMARY,
} = PAYMENTS_MUTATIONS;

describe('EstimatedCostsAndCredits', (): void => {
    beforeEach(() => {
        jest.useFakeTimers("modern");
        jest.setSystemTime(now);
    });
    afterAll(() => {
        jest.useRealTimers();
    });

    it('renders correctly with project and no project usage and charges', async (): Promise<void> => {
        await store.commit(USER_MUTATIONS.SET_USER, user);
        await store.commit(PROJECTS_MUTATIONS.ADD, project);

        const wrapper = mount(EstimatedCostsAndCredits, {
            store,
            localVue,
        });

        await wrapper.setData({ isDataFetching: false });

        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with project and project usage and charges', async (): Promise<void> => {
        await store.commit(PROJECTS_MUTATIONS.ADD, project);
        await store.commit(SET_PROJECT_USAGE_AND_CHARGES, [projectCharge]);
        await store.commit(SET_PRICE_SUMMARY, [projectCharge]);

        const wrapper = mount(EstimatedCostsAndCredits, {
            store,
            localVue,
        });

        await wrapper.setData({ isDataFetching: false });

        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with 2 projects and project usage and charges', async (): Promise<void> => {
        await store.commit(PROJECTS_MUTATIONS.ADD, project);
        await store.commit(PROJECTS_MUTATIONS.ADD, project1);
        await store.commit(SET_PROJECT_USAGE_AND_CHARGES, [projectCharge, projectCharge1]);
        await store.commit(SET_PRICE_SUMMARY, [projectCharge, projectCharge1]);

        const wrapper = mount(EstimatedCostsAndCredits, {
            store,
            localVue,
        });

        await wrapper.setData({ isDataFetching: false });

        expect(wrapper).toMatchSnapshot();
    });
});
