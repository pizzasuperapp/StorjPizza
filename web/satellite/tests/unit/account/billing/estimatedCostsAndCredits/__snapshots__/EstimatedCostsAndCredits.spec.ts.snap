// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EstimatedCostsAndCredits renders correctly with 2 projects and project usage and charges 1`] = `
<div class="current-month-area">
  <h1 class="current-month-area__costs">USD $6.00</h1> <span class="current-month-area__title">Estimated Charges for May 2020</span>
  <p class="current-month-area__info">
    If you still have Storage and Bandwidth remaining in your free tier, you won’t be charged. This information
    is to help you estimate what charges would have been had you graduated to the paid tier.
  </p>
  <div class="current-month-area__content">
    <p class="current-month-area__content__title">DETAILS</p>
    <div class="usage-charges-item-container item">
      <div class="usage-charges-item-container__summary">
        <div class="usage-charges-item-container__summary__name-container"><svg></svg> <span>projectName</span></div> <span class="usage-charges-item-container__summary__amount">
            Estimated Total USD $3.00
        </span>
      </div>
      <!---->
    </div>
    <div class="usage-charges-item-container item">
      <div class="usage-charges-item-container__summary">
        <div class="usage-charges-item-container__summary__name-container"><svg></svg> <span>projectName1</span></div> <span class="usage-charges-item-container__summary__amount">
            Estimated Total USD $3.00
        </span>
      </div>
      <!---->
    </div>
  </div>
</div>
`;

exports[`EstimatedCostsAndCredits renders correctly with project and no project usage and charges 1`] = `
<div class="current-month-area">
  <h1 class="current-month-area__costs">USD $0.00</h1> <span class="current-month-area__title">Estimated Charges for May 2020</span>
  <p class="current-month-area__info">
    If you still have Storage and Bandwidth remaining in your free tier, you won’t be charged. This information
    is to help you estimate what charges would have been had you graduated to the paid tier.
  </p>
  <div class="current-month-area__content">
    <p class="current-month-area__content__title">DETAILS</p>
  </div>
</div>
`;

exports[`EstimatedCostsAndCredits renders correctly with project and project usage and charges 1`] = `
<div class="current-month-area">
  <h1 class="current-month-area__costs">USD $3.00</h1> <span class="current-month-area__title">Estimated Charges for May 2020</span>
  <p class="current-month-area__info">
    If you still have Storage and Bandwidth remaining in your free tier, you won’t be charged. This information
    is to help you estimate what charges would have been had you graduated to the paid tier.
  </p>
  <div class="current-month-area__content">
    <p class="current-month-area__content__title">DETAILS</p>
    <div class="usage-charges-item-container item">
      <div class="usage-charges-item-container__summary">
        <div class="usage-charges-item-container__summary__name-container"><svg></svg> <span>projectName</span></div> <span class="usage-charges-item-container__summary__amount">
            Estimated Total USD $3.00
        </span>
      </div>
      <!---->
    </div>
  </div>
</div>
`;
